src/user/user.service.ts

```diff
@@ -1,29 +1,146 @@
-import { Injectable, NotFoundException } from '@nestjs/common';
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
 import { InjectRepository } from '@nestjs/typeorm';
-import { User } from './entities/user.entity';
-import { Repository } from 'typeorm';
-import { UserDto } from './dto/user';
+import { UNIQUE_USER_EMAIL_CONSTRAINT, User } from './entities/user.entity';
+import { QueryFailedError, Repository } from 'typeorm';
+import { ActivateUserDto, CreateUserDto, UserDto } from './dto/user';
+import { UserAlreadyExists, UserNotActivated } from './user.exceptions';
+import { CryptoService } from '../utils/crypto.service';
+import { UserToken, UserTokenType } from './entities/user-token.entity';
 
 @Injectable()
 export class UserService {
   constructor(
     @InjectRepository(User)
     private readonly userRepository: Repository<User>,
+    @InjectRepository(UserToken)
+    private readonly tokenRepository: Repository<UserToken>,
+    private readonly cryptoService: CryptoService,
   ) {}
 
   async findAll(): Promise<UserDto[]> {
+    //Todo: do we need this route? Filter active/inactive?
     return this.userRepository.find();
   }
 
-  async findOne(username: string): Promise<UserDto> {
+  /**
+   * Returns on user or throws a NotFoundException. If canBeInactive is set, it
+   * also returns a user that has not activated their account yet. Defaults to
+   * false so we always scope this to the active users only and throw a
+   * UserNotActivatedException.
+   * @param email
+   * @param canBeInactive
+   */
+  async findOne(email: string, canBeInactive = false): Promise<UserDto> {
     const user = await this.userRepository.findOne({
-      where: [{ username: username }],
+      where: [{ email: email }],
     });
 
     if (user) {
+      if (!canBeInactive && !user.isActive) {
+        throw new UserNotActivated();
```

Comment:
This could also be a potential security risk as we return the information that such a user exists but it is not activated yet. Could we also throw a not found exception here?

Reply:
Good point. I did not use `NotFound` because I felt that UX-wise, it's nicer to give the user the feedback that they might have to activate their account. But you point is valid, so I might change that.

Reply:
Fixed in #49